w06/test/gad/DualPivotQuicksortTest.java

```diff
@@ -0,0 +1,165 @@
+                        new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30}
+                ),
+                Arguments.of(
+                        new int[]{5, 6, 5, 6, 5, 6, 55, 6, 5, 6, 5, 6, 55, 6, 5, 6, 5, 6, 55, 6, 5, 6, 5, 6, 5},
+                        new int[]{5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 55, 55, 55}
+                ),
+                Arguments.of(
+                        new int[]{30, 25, 20, 15, 14, 10, 5, 1, 0, -5, -20, -100},
+                        new int[]{-100, -20, -5, 0, 1, 5, 10, 14, 15, 20, 25, 30}
+                ),
+                Arguments.of(
+                        new int[]{},
+                        new int[]{}
+
+                ),
+                Arguments.of(
+                        new int[]{5},
+                        new int[]{5}
+                ),
+                Arguments.of(
+                        new int[]{8, 3},
+                        new int[]{3, 8}
+                ),
+                Arguments.of(
+                        new int[]{7, 10},
+                        new int[]{7, 10}
+                ),
+                Arguments.of(
+                        new int[]{4, 2, 7, 2, 9, 4, 1},
+                        new int[]{1, 2, 2, 4, 4, 7, 9}
+                ),
+                Arguments.of(
+                        new int[]{9, 7, 5, 3, 1},
+                        new int[]{1, 3, 5, 7, 9}
+                ),
+                Arguments.of(
+                        new int[]{2, 4, 6, 8, 10},
+                        new int[]{2, 4, 6, 8, 10}
+                ),
+                Arguments.of(
+                        new int[]{9, 3, 7, 1, 5, 2, 6, 8, 4},
+                        new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9}
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    public void dualPivotQuickSortTest(int[] numbers, int[] expected) {
+        int[] numbersCopy = Arrays.copyOf(numbers, numbers.length);
+        DualPivotQuicksort dualPivotQuicksort = new DualPivotQuicksort(DualPivotFinder.getFirstLastPivot(), 0);
+        StudentResult result = new StudentResult();
+
+        dualPivotQuicksort.sort(numbersCopy, result, 0, numbersCopy.length - 1);
+        assertArrayEquals(expected, numbersCopy);
+    }
+
+
+    //for this test to work you have to remove the "result.log()" from your DualPivotQuicksort sort() method
+    @Test
+    public void dualPivotQuickSortTestLarge() {
+        Random random = new Random();
+        int[] numbers = new int[1_000_000];
+        for (int i = 0; i < numbers.length; i++) {
+            numbers[i] = random.nextInt(Integer.MAX_VALUE);
+        }
+        int[] expected = Arrays.copyOf(numbers, numbers.length);
+
+        DualPivotQuicksort dualPivotQuicksort = new DualPivotQuicksort(DualPivotFinder.getRandomPivot(10), 0);
+        StudentResult result = new StudentResult();
```

Comment:
Here it's the same as in Line `157`. Especially in this Test the logging is rising the runtime and this caused that some devices might not finish this test.